File: pair_socks/pair_socks.ts
function sockMerchant(arr) {
  let temp = {};
  let counter: number = 0;

  for (let value of arr) {
    temp[value] = ++temp[value] || 1;
  }

  for (let value in temp) {
    if (temp[value] % 2 !== 0 && temp[value] > 1) {
      counter += (temp[value] - (temp[value] % 2)) / 2;
    }

    if (temp[value] % 2 === 0) {
      counter += temp[value] / 2;
    }
  }

  return counter;
}

console.log(sockMerchant([10, 20, 20, 10, 10, 30, 50, 1020]));
console.log(sockMerchant([10, 20, 20, 10, 10, 10, 30, 50, 50, 50, 1020]));

File: pair_socks/pair_socks.py
def socksMerchant(arr):
    temp = {}
    counter = 0
    
    for i in range(len(arr)):
        temp[arr[i]] = temp.get(arr[i], 0) + 1
        
        
    for i in temp:
        if temp[i] % 2 != 0 and temp[i] > 1:
            counter += (temp[i] - (temp[i] % 2)) / 2 
            
        if temp[i] % 2 == 0:
            counter += temp[i] / 2
    
    return counter
    
print(socksMerchant([10, 20, 20, 10, 10, 10, 30, 50, 1020]))
